### DISCUSSION TITLE
Extending Mantine Default Size & Spacing Values

### DISCUSSION THREAD

**Discussion by KMathisGit** [mantinedev/mantine#4966] [Feature requests]
I have been struggling extending mantine theme to support additional size and spacing tokens such as `2xl`. 

I realize that we can add additional values by adding them via `createTheme` and passing this to the `MantineProvider` for instance:
```
<MantineProvider theme={createTheme({ spacing: { '2xl': '3rem' } }) }> 
  <App />
</MantineProvider>
```

This does work however this does not extend the MantineSpacing type definition and leads to bad DX as intellisense and suggestions will only pick up the default MantineSpacing values `'xs' | 'sm' | 'md' | 'lg' | 'xl'`

I could simply override the MantineTheme type definition with my own declaration but I would rather not do so for numerous reasons...

I feel like there should be some way to add additional tokens like I described above. Or am I overlooking some already existing behavior?

**Comment by JunlinPan-Crypto** [comment]
Same here.
Mantine only provide the declaration for `theme.other` and `theme.colors` overriding - https://mantine.dev/guides/typescript/#theme-object-declarations

**Comment by DavideBruner** [comment]
In case you're still facing this issue, here's what worked for me. I extended the Mantine theme to include the 'xxl' spacing token without overriding the core type definition.

First, in your global.d.ts:
```ts
import { MantineTheme } from "@mantine/core";

declare module "@mantine/core" {
  export interface MantineThemeOverride extends MantineTheme {
    spacing: Record<"xs" | "sm" | "md" | "lg" | "xl" | "xxl", string>;
  }
}

```

Then, in your theme overrides:
```ts
spacing: {
    xxl: "4rem",
},
...
```

**Reply by enyelsequeira** [reply]
> In case you're still facing this issue, here's what worked for me. I extended the Mantine theme to include the 'xxl' spacing token without overriding the core type definition.
> 
> First, in your global.d.ts:
> 
> ```ts
> import { MantineTheme } from "@mantine/core";
> 
> declare module "@mantine/core" {
>   export interface MantineThemeOverride extends MantineTheme {
>     spacing: Record<"xs" | "sm" | "md" | "lg" | "xl" | "xxl", string>;
>   }
> }
> ```
> 
> Then, in your theme overrides:
> 
> ```ts
> spacing: {
>     xxl: "4rem",
> },
> ...
> ```

![image](https://github.com/user-attachments/assets/8045505c-1e13-4d3e-b53a-9049ede93232) 

Hey @DavideBruner  I am wondering if this actually worked for you, I've tried doing what you did and does not work for me wondering if you had a demo where that worked? also wondering if by doing that were you able to get autocomplete on the props as well?

**Reply by jrozbicki** [reply]
This doesn't work as `MantineThemeOverride` is a `type`, not `interface

**Reply by DavideBruner** [reply]
Hey everyone, sorry for missing the tags and the conversation here! I just caught up with the thread and wanted to clarify a few things.

I have to admit I haven’t been super active in the Mantine repo lately since I’ve been switching over to Radix UI / shadcn/ui for newer projects. Also, GitHub notifications can be tricky to keep up with 😅.

To address the original question—yes, the approach I shared was a bit of a workaround rather than a fully supported extension method. At the time (Mantine v6.0.x), the `MantineSize` type allowed both predefined sizes and arbitrary strings, which let us extend spacing tokens without TypeScript errors. However, it didn't provide proper autocomplete, which was definitely a downside.
<img width="590" alt="Screenshot 2025-03-24 at 15 32 00" src="https://github.com/user-attachments/assets/e580a45a-30d3-4a98-89e4-9a162dad9982" />


I see the struggle with getting proper DX improvements here, and I agree it would be great if Mantine supported module augmentation for spacing in a way similar to `MantineThemeColorsOverride`.

I don’t have a public repo to share at the moment, but I can provide some screenshots and details about the version I was using if that helps. These projects were built in late 2022 and were using Mantine v6.0.x.

**Comment by jrozbicki** [comment]
I'd love a way to improve DX and extend e.g. `spacing` autocomplete with `2xl`. 

It does work in runtime when you add it in `createTheme({})` and later use in `<Stack gap="2xl">`.

Would be awesome to leverage module augmentation similarly to how we can do with `MantineThemeColorsOverride` for `colors`.

**Reply by rtivital** [reply]
It is hard to implement unfortunately

**Reply by wsfuller** [reply]
In similar situation where I need a smaller spacing attribute `xxs`. Yes the theme can be expanded but there's no auto complete so it's introducing institutional knowledge, not ideal.

I was trying to do the following with several variations but was never able to get this working

```
mantine.d.ts

import '@mantine/core';

declare module '@mantine/core' {
  export interface MantineThemeSpacing {
    xxs: number;
  }

  export interface MantineTheme {
    spacing: MantineThemeSpacing;
  }
}

``` 

```
tsconfig.json

{
  "compilerOptions": {...},
  "include": [
    ...,
    mantine.d.ts
  ],
  ...
}
```

```
theme.ts

const theme = createTheme({
  ...,
  other: {
    tokens: { ... },
  },
  components: {...},
  spacing: {
    xxs: ...
  }
});
```

**Comment by doelgonzo** [comment]
Watching this, as I also need to add a 2xl and 2xs option

**Reply by Merces-dev** [reply]
Hi @doelgonzo ! I'm using `@mantine/core@^8.0.1`, and I was able to extend the theme types successfully by following these steps:

1. **Created a custom `mantine.d.ts` file** to override the theme types:

```ts
// src/types/mantine.d.ts
import '@mantine/core';
import { fontSizes, radius, spacing } from '@/configs/mantine/theme';

declare module '@mantine/core' {
  export interface MantineThemeSizesOverride {
    spacing: typeof spacing;
    fontSizes: typeof fontSizes;
    radius: typeof radius;
  }
}
```

2. **Defined the constants** in my custom theme configuration file (`@/configs/mantine/theme`):

```ts
// @/configs/mantine/theme

import { createTheme } from '@mantine/core';

export const fontSizes = {
  xl: '24px',
  lg: '20px',
  md: '16px',
  sm: '14px',
  xs: '12px',
} as const;

export const spacing = {
  none: '0px',
  '4xs': '2px',
  xxxs: '4px',
  xxs: '8px',
  xs: '12px',
  s: '16px',
  m: '24px',
  l: '32px',
  xl: '48px',
  xxl: '64px',
  xxxl: '80px',
  '4xl': '96px',
} as const;

export const radius = {
  none: '0px',
  '4xs': '2px',
  xxxs: '4px',
  xxs: '8px',
  xs: '12px',
  s: '16px',
  m: '24px',
  l: '32px',
  xl: '48px',
  full: '1000px',
} as const;

export const theme = createTheme({
  primaryColor: 'red',
  primaryShade: 6,
  fontSizes,
  radius,
  spacing,
});
```

3. **Included the type declaration in `tsconfig.json`** so TypeScript picks it up:

```ts
//tsconfig.json

{
  "include": [
    "src/types/**/*.d.ts"
  ]
}
```
With this setup, I was able to access all custom theme values (fontSizes, spacing, radius) with full IntelliSense support in the code editor.

This setup allowed me to safely customize Mantine’s theme. Hope this helps someone else trying to do the same!

**Reply by codex-diem** [reply]
No way for v7? 
Avoiding migration to v8?
